doc/design/embargo-redesign.md

```diff
@@ -116,3 +116,49 @@ sequenceDiagram
 
 - Deleting tags of 10,000 objects took ~18 seconds
 - Deleting tags of 100,000 objects took ~2 minutes
+
+## Semantic Quirks of Embargo
+
+During redesign deliberations, we noticed some oddities of how embargo might interact with deduplication in ways that raise some questions about the intended semantics of embargo. This section reviews some examples and states those questions for discussion and resolution among the full team.
+
+### Embargoed Dandiset includes open asset
+
+**Scenario.** Someone creates an open Dandiset 1, uploading asset A to it. They then create an embargoed Dandiset 2, also uploading asset A to that.
+
+**Quirk.** Dandiset 2, though embargoed, contains an asset that is open and therefore world-readable. In spirit this contradicts the embargo on Dandiset 2.
+
+**Remediation.** None needed, despite the apparent contradiction. Though Dandiset 2 contains a world-readable asset, no one who is not authorized to view the contents of Dandiset 2 knows that fact; for all they know, any data at all, *including open assets stored in the archive*, could be in that Dandiset. Thus, no excess information about Dandiset 2 is leaked.
+
+### Open Dandiset includes embargoed asset
+
+**Scenario.** Roughly the opposite of the last scenario: someone creates an *embargoed* Dandiset 3, uploading asset B to it. They then create an *open* Dandiset 4, also uploading B to that.
+
+**Quirk.** Dandiset 4 now contains a hidden asset, which contradicts the openness of Dandiset 4 and seems to compromise the secrecy of Dandiset 3.
```

Comment:
after re-reading original design above I noted that although it describes technical aspect of the solution on S3 level, it does not cover at all annotation of embargoed "stuff" within archive DB, and might be worth clarifying on what  exactly (dandiset? asset? blob? zarr? all of that) and how (metadata? extra DB column? both) is annotated to be either open or embargoed?  I guess intention is to just keep it as is now (EmbargoedAssetBlob, EmbargoedZarrArchive + metadata annotation of embargoed dandiset), but feels like a missing detail here which instructs implementation logic and potentially these use-cases / quirks.

Reply:
I sort of agree with you, but I see it a little bit differently: I don't want to formalize what it means for "assets to be embargoed", because "embargo" is a very specific application-level concept that models *NIH-authorized embargo* for a very specific situation. In that sense, how we are keeping such embargoed Dandisets' contents hidden from public view is an implementation detail, and I think it would actually be good to use a different term than "embargoed" to describe the state of such assets.

Where we agree (I believe) is the need to formalize what "embargo" means (in my opinion, it is simply the state of a Dandiset which includes its contents and existence being hidden from unauthorized users).

As for implementation: yes, I think we are planning to keep the current models in place, only changing them if (1) the changes improve the code and (2) the requirements/behavior of the system are not changed. (This is essentially what it means to be an "implementation detail".)

Reply:
ref: https://github.com/dandi/dandi-archive/issues/1787 a somewhat related since details of implementation (either to be a part of this design doc or not) and e.g. interaction with metadata remain not formalized in a design doc, or is it?

Reply:
>  it is simply the state of a Dandiset which includes its contents and existence being hidden from unauthorized user

i would strike "and existence". it does not matter from an embargo perspective whether dandiset 2024 exists or not. simply knowing that no content about it is exposed is sufficient.

embargo/restricted - only certain authorized individuals/entities have access to that resource (whatever the resource may be).

perhaps a few additional things to note:
- asset ids and blobs are decoupled
- blob ids and blob checksums are decoupled and currently fall under deduplication principles rather than embargo principles. 

thus from an implementation perspective:
- a user should be notified if a blob being uploaded to an embargoed dandiset already exists in public view (could easily be possible for a host of reasons). i believe the CLI doesn't allow 0 length files. i'm not sure the API has any restriction. 
- one needs to consider as @yarikoptic wrote as to what happens when these two worlds collide. same blob uploaded to two different embargoed dandisets (also completely possible). current deduplication rules would not allow a different id. 

both `embargo` (being discussed here) and `restricted` (to be implemented soon) are resource specific concepts. as such i would really bring rebac into the picture. the resources in consideration are dandisets, dandiset metadata, assets, and blobs. whether we allow access control on all of these elements is a policy decision, but implementation wise we should consider them as separate resources. dandisets can contain mixtures of access assets for whatever reason and several users have asked for asset level granularity for access for different use cases. a lot of the considerations in quirks would be covered under rebac, which turns things into policy decisions for access.

Reply:
Could you expand the acronym "rebac"? (I have seen it stand for different things and I want to disambiguate what you are referring to.)

> dandisets can contain mixtures of access assets for whatever reason and several users have asked for asset level granularity for access for different use cases.

I was really hoping to avoid this type of generality, especially without hard requirements on how this would work. My preference is to work from well-defined concepts such as "NIH embargo" or "restricted access" (the latter is not super well defined but it is at least a domain-level concept) in order to deliver those chunks of features to users.

I understand that users have asked for the fully general idea of mixing access levels of assets, but if we don't understand the actual software requirements here (inlcuding some of the use cases those users are trying to implement), this is a perfect recipe for half-baked implementations plus subsequent rework that we have been desperately trying to avoid.

I am happy to discuss this direction further, but in the current design I would like to pin down how embargo should work in the face of these "quirks" (and some shadow of future capabilities we want to develop). I think we are well on the road to accomplishing this, but I wanted to draw this boundary so that we don't get bogged down in trying to understand three things at once (i.e., embargo, restricted, general asset access).

Reply:
rebac --> resource based access control

regarding use cases:

- ML leaderboards are what has driven most of the user requests on granular access, specifically to create training, validation, and multiple test data sets. the current solution is simply to have multiple dandisets.  
- Embargoed master data with periodic releases of subsets of data associated with papers. again doable from the current setup, not efficient or easy to do from a user perspective.

> so that we don't get bogged down in trying to understand three things at once (i.e., embargo, restricted, general asset access).

i'm fine with not getting bogged down, but if the entire thing will require a complete overhaul again when we implement restricted/general asset access, i would rather have that be discussed in this context.

Reply:
> i'm fine with not getting bogged down, but if the entire thing will require a complete overhaul again when we implement restricted/general asset access, i would rather have that be discussed in this context.

This is a fair position to take. FWIW, my intuition is that with the ability to hide/open assets already existing, we should be able to build out embargo, and "restricted assets", and the more general mixed-access Dandiset concepts.

For the latter two (which don't exist yet) we still have room to understand requirements and connect them to the existing capabilities of the system--I simply want to make sure we develop that full understanding before anything else (which includes discussions now about those requirements, etc.).

Reply:
BTW -- isn't "EMBARGOED" is just a "RESTRICTED" with a dandiset-level rule for when "RESTRICTED" could be automatically be converted to "OPEN"?

Reply:
> isn't "EMBARGOED" is just a "RESTRICTED" with a dandiset-level rule

yup, just policies are different as to how access is granted. embargoed currently restricted to owners. for restricted, we will need both owners and viewers who have access. hence the consideration of rebac (resource based) + rbac (role based).

Reply:
I think model could be simplified if EMBARGOED would just become RESTRICTED with the policy for when it to become OPEN, and otherwise to have the same permission models as RESTRICTED, i.e. to have owners and viewers, where the viewers have no authority to change it, e.g. to make it OPEN through the .  I think it might be one of the possible ways collaborators on EMBARGOED dataset could leak it: through that [Explicit "make content of a dandiset"](https://github.com/dandi/dandi-archive/pull/1802#discussion_r1440043850) entry point I was talking about.  Sure thing they would still be able to make it OPEN through leaking via uploading them as OPEN to another OPEN dataset (unless we forbid such an action as @waxlamp excercised in [this remediation](https://github.com/dandi/dandi-archive/pull/1802/files#diff-f3d72d358571f346f0a3e26e3bf309127d6af6aac1e39f60e291b947d86eeb19R140), but I think it would be potentially causing other issues (pretty much a "race condition" on some possibly trivial content as e.g. a json file with just `[]` or alike)

Reply:
I really do understand the pull to "simplify" the model so that embargo (which we really need to fix), "restricted", and a more general "mixed access" all appear to become simple extensions thereof, but doing so will actually make it harder to deliver the functionality as envisioned, unless we *write down requirements for each* and then validate that the implementation plan will support all requirements.

I do feel like we're on the verge of figuring out the broad semantics of "restriction" that will support everything, but I am really feeling the exposure to the risk of getting it wrong (and thus triggering another round of discussions and then rework to fix a broken attempted implementation). I'm going to schedule a meeting so we can discuss this synchronously and try to eliminate those risks.